File: java/src/TreeNode.java
import java.util.ArrayList;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    ArrayList preorder() {
        return preorder(new ArrayList());
    }

    ArrayList preorder(ArrayList order) {
        order.add(val);
        if (null != left) order.addAll(left.preorder());
        if (null != right) order.addAll(right.preorder());
        return order;
    }

    ArrayList inorder() {
        return inorder(new ArrayList());
    }

    ArrayList inorder(ArrayList order) {
        if (null != left) order.addAll(left.inorder());
        order.add(val);
        if (null != right) order.addAll(right.inorder());
        return order;
    }
}

File: go/tree_node.go
package main

import "fmt"

// TreeNode is just tree node
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preOrderP(n *TreeNode) {
	if n == nil {
		return
	}
	arr := []int{}
	preOrderCollect(n, &arr)
	fmt.Println(arr)
}

func preOrderCollect(n *TreeNode, arr *[]int) {
	if n == nil {
		return
	}
	*arr = append(*arr, n.Val)
	preOrderCollect(n.Left, arr)
	preOrderCollect(n.Right, arr)
}

func inOrderP(n *TreeNode) {
	if n == nil {
		return
	}

	values := []int{}
	queue := []*TreeNode{n}
	for {
		if len(queue) == 0 {
			break
		}

		curr := queue[0]
		queue = queue[1:]
		values = append(values, curr.Val)

		if curr.Left != nil {
			queue = append(queue, curr.Left)
		}
		if curr.Right != nil {
			queue = append(queue, curr.Right)
		}
	}
	fmt.Println(values)
}
